framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -272,106 +273,206 @@ protected boolean areEqualInHierarchy(
+            areEqualVisitHistory.put(inside, outside, currentTop, result);
+            return result;
         }
-        Boolean previousResult = areEqualVisitHistory.get(inside, outside, currentTop);
-        if (previousResult != null) {
-            return previousResult;
+        if (canBeCovariant) {
+            return isSubtype(inside, outside, currentTop);
         }
-
-        areEqualVisitHistory.put(inside, outside, currentTop, true);
-        boolean result =
-                isContainedWildcard(
-                        inside, outside, outsideUpperBound, outsideLowerBound, canBeCovariant);
-        areEqualVisitHistory.put(inside, outside, currentTop, result);
-        return result;
+        return areEqualInHierarchy(inside, outside);
     }
 
-    private boolean isContainedWildcard(
-            AnnotatedTypeMirror inside,
-            AnnotatedTypeMirror outside,
-            AnnotatedTypeMirror outsideUpperBound,
-            AnnotatedTypeMirror outsideLowerBound,
-            boolean canBeCovariant) {
-
-        if (inside.equals(outside)) {
-            // If they are equal, outside always contains inside.
-            return true;
-        }
-
-        if (inside.getKind() == TypeKind.WILDCARD) {
-            outsideUpperBound =
-                    checker.getTypeFactory()
-                            .widenToUpperBound(outsideUpperBound, (AnnotatedWildcardType) inside);
-        }
-        while (outsideUpperBound.getKind() == TypeKind.WILDCARD) {
-            if (ignoreUninferredTypeArgument(outsideUpperBound)) {
-                return true;
+    /**
+     * Returns true if {@code outside} contains {@code inside}, that is, if the set of types denoted
+     * by {@code outside} is a superset of or equal to the set of types denoted by {@code inside}.
+     * See {@link #isContainedBy(AnnotatedTypeMirror, AnnotatedTypeMirror, boolean)} for a full
+     * explanation.
+     *
+     * <p>Roughly, the algorithm works as follows (assuming {@code canBeCovariant} is false):
+     *
+     * <ul>
+     *   <li>If inside is a bound type, recur on the explicit bound.
+     *   <li>If one of outside's bounds is itself a bound type, recur on that bound.
+     *   <li>Otherwise, return {@code outside.lower <: inside && inside <: outside.upper}.
+     * </ul>
+     *
+     * @param inside a possibly-contained type
+     * @param outside a possibly-containing type
+     * @param canBeCovariant whether or not type arguments are allowed to be covariant
+     * @return true if inside is contained by outside, or if canBeCovariant == true and {@code
+     *     inside <: outside}
+     */
+    protected boolean isContainedByBoundType(
+            AnnotatedTypeMirror inside, BoundType outside, boolean canBeCovariant) {
+        if (BoundType.isBoundType(inside)) {
+            BoundType insideBoundType = new BoundType(inside);
+            if (!insideBoundType.hasExplicitLowerBound && !outside.hasExplicitLowerBound) {
```

Comment:
Rather than using negation, it might be a touch clearer here to add a method hasExplicitUpperBound() to `BoundType`.  Then this test can be
```java
if (insideBoundType.hasExplicitUpperBound() && outside.hasExplicitUpperBound())
```
which I think is more readable.